builtin/providers/aws/mutex.go

```diff
@@ -0,0 +1,49 @@
+package aws
+
+import (
+	"log"
+	"sync"
+)
+
+// MutexKV is a simple key/value store for arbitrary mutexes. It can be used to
+// serialize changes across arbitrary collaborators that share knowledge of the
+// keys they must serialize on.
+//
+// The initial use case is to let aws_security_group_rule resources serialize
+// their access to individual security groups based on SG ID.
+type MutexKV struct {
+	sync.Mutex
+	store map[string]*sync.Mutex
+	once  sync.Once
+}
+
+func (m *MutexKV) init() {
+	m.store = make(map[string]*sync.Mutex)
+}
+
+// Returns a mutex for the given key, no guarantee of its lock status
+func (m *MutexKV) Get(key string) *sync.Mutex {
+	m.once.Do(m.init)
```

Comment:
This looks like it probably works fine - I might have just done like a `NewMutexKV()` and used it in the initializer for the global. Not sure if there is any downside in using this at `Get()` time.